libvast/src/data.cpp

```diff
@@ -285,4 +338,108 @@ bool convert(const data& d, json& j, const type& t) {
   return convert(d, j);
 }
 
+namespace {
+
+data parse(const YAML::Node& node) {
+  switch (node.Type()) {
+    case YAML::NodeType::Undefined:
+    case YAML::NodeType::Null:
+      return data{};
+    case YAML::NodeType::Scalar: {
+      auto str = node.as<std::string>();
+      data result;
+      // Attempt some type inference.
+      if (parsers::boolean(str, result))
+        return result;
+      // Attempt maximum type inference.
+      if (parsers::data(str, result))
+        return result;
+      // Take the input as-is if nothing worked.
+      return str;
+    }
+    case YAML::NodeType::Sequence: {
+      list xs;
+      xs.reserve(node.size());
+      for (size_t i = 0; i < node.size(); ++i)
+        xs.push_back(parse(node[i]));
+      return xs;
+    }
+    case YAML::NodeType::Map: {
+      record xs;
+      xs.reserve(node.size());
+      for (auto& pair : node)
+        xs.emplace(pair.first.as<std::string>(), parse(pair.second));
+      return xs;
+    }
+  }
+  return {}; // makes GCC happy
+}
+
+} // namespace
+
+caf::expected<data> from_yaml(std::string_view str) {
+  try {
+    // Maybe one glory day in the future it will be possible to perform a
+    // single pass over the input without incurring a copy.
+    auto node = YAML::Load(std::string{str});
+    return parse(node);
+  } catch (const YAML::Exception& e) {
+    return caf::make_error(ec::parse_error,
+                           "failed to parse YAML at line/col/pos", e.mark.line,
+                           e.mark.column, e.mark.pos);
+  }
+}
+
+namespace {
+
+void print(YAML::Emitter& out, const data& x) {
+  // clang-format off
+  auto f = detail::overload(
+    [&out](const auto& x) { out << to_string(x); },
```

Comment:
Can we have this function do `static_assert(std::always_false_v<decltype(x)>, "unsupported type")` and instead list all types explicitly? Otherwise this becomes easy to overlook in future refactorings.

Reply:
I'll provide overloads for all types and ditch `auto`. Then compiler will complain if an overload is missing.

Reply:
Good call.